### DISCUSSION TITLE
Documentation tg file format

### DISCUSSION THREAD

**Discussion by profOnno** [helge17/tuxguitar#264] [General]
Hello, 

Just trying out TuxGuitar, and i like it.

I also like the Godot game development platform. So now i would like to use the tuxguitar file format directly. I was looking for some documentation on the file format but couldn't find it. Is there documentation on the tg file format? 
Or should I follow the following locations, to figure it out from there?

https://github.com/helge17/tuxguitar/blob/master/common/TuxGuitar-lib/src/org/herac/tuxguitar/io/tg/TGSongReaderImpl.java
https://github.com/helge17/tuxguitar/blob/master/common/TuxGuitar-lib/src/org/herac/tuxguitar/io/tg/TGStream.java

Kind regards,

Menno

**Comment by guiv42** [comment]
Hi

up to now TuxGuitar uses a proprietary binary format, which as far as I know is not documented. I think the java code you pointed to is the right one.

I don't know what you want to do with tg files, but TuxGuitar can also export files under different formats (see [application help](https://www.tuxguitar.app/files/1.6.1/desktop/help/file_formats.html)). Some formats may be easier to use, but some information may be lost.
Note: if you want to use MusicXML format I advise to use [latest snapshot](https://github.com/helge17/tuxguitar/releases), export to this file format has been re-implemented.

Note: the question of tg file format is currently open. We will need to modify this format to integrate new features.
That should be a good opportunity to switch to a format easier to use, and to document it. This is not precisely scheduled yet, but you can expect such an evolution in the coming months.

**Reply by profOnno** [reply]
Thanks for the quick response,

Should i expect information loss with the MusicXML?  I looked at the reference and it looks like everything is there, lyrics, bents hammer on / hammer off.

About the "new" tg file format, do you already have some ideas?
I was looking at this site:
https://www.sitepoint.com/data-serialization-comparison-json-yaml-bson-messagepack/
Messagepack seems promising. I'm not really familiar with code base of TuxGuitar, so I don't know the problems / challenges  you're facing. So maybe I'm totally off with my comment.

I wish you the best, with the updates. And I'll give the MusicXML a try, after update-ing to the latest snapshot :p

Cheerz,

Menno

**Reply by guiv42** [reply]
> Should i expect information loss with the MusicXML?

It seems MusicXML export was developed quite a long time ago. I updated it very recently so that it's now compliant with last version of xml schema (4.0). But data export is still quite basic, so yes you shall definitely expect some significant information loss (typically lyrics, bends, hammer-on/pull-off...). That still needs to be improved.

And thanks for the link about serialization. No idea currently of any solution for TuxGuitar (it's a bit early), but we will have a look at it.

**Reply by profOnno** [reply]
Here is some rust code that uses a serialization framework: 
```rs

#[allow(non_camel_case_types)]
#[derive(Deserialize, Serialize, PartialEq, Debug)]
enum Software {
    TuxGuitar_1_6_1,
    TuxGuitar_1_2_1,
}

#[derive(Deserialize, Serialize, PartialEq, Debug)]
struct Encoding {
    software: Software,
}

#[derive(Deserialize, Serialize, PartialEq, Debug)]
struct Creator {
    #[serde(rename = "$text")]
    creator: String,
    #[serde(rename = "@type")]
    composer: String, // should be enum?
}

#[derive(Deserialize, Serialize, PartialEq, Debug)]
struct Identification {
    creator: Creator,
    encoding: Encoding,
}

fn main() {
    let ident = Identification {
        creator: Creator {
            creator: "John".to_string(),
            composer: "composer".to_string(),
        },
        encoding: Encoding {
            software: Software::TuxGuitar_1_6_1,
        },
    };

    let xml: String = se::to_string(&ident).unwrap();

    println!("{}", xml);
    let mut file = File::create("output.xml").unwrap();
    write!(file, "{}", xml);
    println!("{}", serde_json::to_string(&ident).unwrap());
}
```
Output:
```
<Identification><creator type="composer">John</creator><encoding><software>TuxGuitar_1_6_1</software></encoding></Identification>
{"creator":{"$text":"John","@type":"composer"},"encoding":{"software":"TuxGuitar_1_6_1"}}
```
So the struct is kind of class. The derive is something like a python decorator. The underlying serialization framework is `serde`. The derive adds code that turns the struct into "tokens" for serde. Then there are frameworks to convert the serde "tokens" to the desired output, like json and xml.

Afaik, Java also has the Serialization pattern.
https://www.baeldung.com/java-serialization
https://www.baeldung.com/java-serialization-approaches
https://stackabuse.com/serialize-and-deserialize-xml-in-java-with-jackson/

Not sure if this is usefull information, if not, ignore :p

Menno

**Comment by guiv42** [comment]
Evolution of file format is needed by several pending feature requests. So, let's re-open this discussion : what file format for next TuxGuitar release ?

Just a few personal thoughts about this.

**Mandatory requirements**
- new file format shall be *implementable* and *maintainable* by current (reduced) development team, made of non-professionals only. This is a critical requirement, that could lead us to a technically sub-optimal choice.
- TuxGuitar shall be able to open all old .tg files
- new file format shall include additional data
- trying to open a new file with an old TuxGuitar shall fail smoothly, with an explicit message to user suggesting app upgrade: no crash, no "coded" error message that only developers can understand

**Nice-to-have requirement**
- new file format shall be usable by third party software: either well documented or self-explanatory (or both)

**Open points**
- I don't know if it's worth warning user about incompatibility with old versions. Previous file format modifications were implemented without this precaution.
- should we change file extension, e.g. using ".tg2"?
- I don't know if upstream compatibility should be considered. Making one TG version able to open files created by future versions sounds seducing, but it's also risky. Don't forget there is no test infrastructure in current code base.
The simple strategy used up to now proved efficient: don't try to open newer formats, just ask user to upgrade.

Also, care shall be taken before introducing dependencies to external libs, and the maintenance costs it could create in the future (see vst, gervill, or even recent issues with SWT). We cannot afford a dependency issue for such a core feature.

Considering this:
- manually built binary is still a credible candidate I think
- why not compressed XML or JSON? (warning: dependencies)
- protobuf? (same warning)
- take care about serialization librairies. Some TuxGuitar data structures could create infinite recursive loops : e.g. a TGTrack instance contains a list of TGMeasure instances, and each TGMeasure contains a reference to its parent TGTrack

**Documented** suggestions are welcome.

**Reply by belfie13** [reply]
i think if the format was going to change then make it future proof. so any syntax it doesn't understand is probably a new feature and it should just be skipped. like how css just drops any declarations it doesn't know and resumes parsing after the next statement terminator.

we should write a feature list for including in the file format? do we have one somewhere?

**Reply by guiv42** [reply]
> i think if the format was going to change then make it future proof. so any syntax it doesn't understand is probably a new feature and it should just be skipped. like how css just drops any declarations it doesn't know and resumes parsing after the next statement terminator.

It's not so obvious... E.g. if we design TuxGuitar version N so that it should be able to read files produced by version N+1 (discarding new data), it means that before releasing N+1 we **must** check that it works. And then, when version N+10 is released, we **must** check behavior of all versions N to N+9. This is why I'm not enthusiastic: we don't have any test infrastructure in place to do all these tests automatically. Manually? We will not be able to test everything.
Today's situation is far less risky: when version N tries to open a file, and doesn't recognize it's been produced by version N or precedent, then it does not even try to open it. Less ambitious for sure, but far less risky (development resources are *very* limited and already busy!)

> we should write a feature list for including in the file format? do we have one somewhere?

By definition such a list would be unstable, considering possible new feature requests can arrive any time. I identified feature requests needing file format evolution end of April in [this post](https://github.com/helge17/tuxguitar/discussions/41#discussioncomment-9266444)

**Reply by belfie13** [reply]
Without much experience in this environment, I'll try not to take too much of your time.

> > i think if the format was going to change then make it future proof.
> if we design TuxGuitar version N so that it should be able to read files produced by version N+1 (discarding new data), ... when version N+10 is released, we **must** check behavior of all versions N to N+9.

yes unit testing is painful and crashing should be avoided. I thought there might be an easier way.

> > we should write a feature list for including in the file format? do we have one somewhere?
> By definition such a list would be unstable, considering possible new feature requests can arrive any time.

yes but I was referring to a document describing the final file format that only gets updated when something changes the way files are created. feature requests would not be part of that until they are committed.

thanks, great feedback

i'll be of more help one day, soon as i get my head around it

**Comment by guiv42** [comment]
So, finally I've gone for compressed xml: format is already used by TuxGuitar (.gpx, .musicxml) and it checks all choice criteria above.

A first version without compression is available in my [dev branch](https://github.com/guiv42/tuxguitar/tree/fileFormat). Currently it's a plugin, only built in Linux/SWT version. It uses `.xml` extension.
Feedback would really be appreciated, especially on the [xsd schema](https://github.com/guiv42/tuxguitar/blob/fileFormat/desktop/dev-fileformat/share/tuxguitar.xsd)

Automated testing is in place, and proved *very* useful to guarantee full compatibility. Thanks @profOnno for your suggestions in #303.
I have [tested](https://github.com/guiv42/tuxguitar/blob/2be8b89f99c628bdda79d2556976de397d6eca86/desktop/dev-fileformat/src/test/java/org/herac/tuxguitar/io/devfileformat/TestFileFormat.java#L51) the file format is evolutive.

Still todo, optimize file size: add compression, remove xml indentation (?)

**Comment by guiv42** [comment]
Added compression, still in my [dev branch](https://github.com/guiv42/tuxguitar/tree/fileFormat)
some performance optimization to come, currently it can be quite slow to open large files

**Comment by guiv42** [comment]
update, in new [dev branch](https://github.com/guiv42/tuxguitar/tree/fileFormat2)
performance optimization done, and integration as default file format also

**Reply by Osmiogrzesznik** [reply]
Link is unavailable , 404.

**Reply by guiv42** [reply]
It was merged into [tuxguitar-next](https://github.com/helge17/tuxguitar/blob/tuxguitar-next/common/TuxGuitar-lib/src/test/resources/tuxguitar_20.xsd) branch 
Warning, it's not fully stable yet.
